### ISSUE TITLE
Multiple Machines Serving Golem Behind One Router On One LAN

### ISSUE THREAD

**Issue by ThomIves** [golemfactory/clay#2837]
Golem, I am enjoying being part of your extended team already. One of many growth items needed is the following: a clear procedure with all necessary details to setup multiple machines that exist on the same LAN behind the same router. I did get some direction from the Golem Chat Tech channel, but it's simply not enough. I need more detail about what to do in the config file. I am personally on Linux Mint. I have one of my servers pointed to golem successfully. I am trying to point two more servers at it, but they do not seem to want to connect. I believe all that I need to do is make changes to my config files on each machine and to add port forwarding with the proper advancement of port numbers on my router. Could someone from the provide clear step by step instructions to accomplish this? Thanks, Thom

**Comment by maaktweluit** [comment]
Hey @ThomIves !

There are 2 answers to your question:
- Enable Upnp on your router ( needs to match Upnp type in golem )
- Configure ports manually on your golem nodes

To configure the ports on your golem nodes:
- update 2 ports for golem
- update 1 port for hyperg

for golem update the `start_port` setting in the `./.local/share/golem/default/mainnet/app_cfg.ini` file to the first port you want to use. It is recommended to increment the `start_port` between nodes in steps of 2 since golem uses 2 ports.

then we need to start hyperg on a new port. grab the arguments from a running golem setup with `ps -aux | grep hyperg`. Then replace or append the argument `--port 3283` ( this is the default 3282 + 1 ). it will look something like: `hyperg -d ~/.local/share/golem/... --port 3283`. Run this in a separate terminal and keep it open while running golem.

After saving the config for golem and running hyperg with the right port you are ready to start golem and it will connect on the new ports.

Only some port forwarding left and you are ready to compute!

Please let me know if you have any more questions about this, you can also tag me under the same username on the golem chat :)

**Comment by ThomIves** [comment]
@Jet86 taught me how to setup the router for multiple machines. Do the
settings in the attached image look good for machines 1, 2, and 3.


​


Thom Ives, Call or Text 208.391.3585
Realtor, Businessman, Engineer

**Comment by ThomIves** [comment]
Also, I posted your answer on the golem tech chat. I have not received
consistent answers on this - not angry - just wanting to help get this
clear for myself and others. We still need a step by step procedure with
adequate detail for everyone. Also, it would be nice if the app had notes
on this during startup. Thanks, Thom


Thom Ives, Call or Text 208.391.3585
Realtor, Businessman, Engineer

**Comment by ThomIves** [comment]
I guess we are still not there. The output for the hyperg command gave the
following:

thom@thom-R710 ~ $ ps -aux | grep hyperg
thom     20252  0.4  0.4 1457052 201684 pts/0  Sl+  May28  16:21 hyperg
--db /home/thom/.local/share/golem/default/mainnet/hyperg --logfile
/home/thom/.local/share/golem/default/mainnet/logs/hyperg.log
thom     31760  0.0  0.0  14224  1012 pts/1    S+   14:19   0:00 grep
--color=auto hyperg
thom@thom-R710 ~ $



Thom Ives, Call or Text 208.391.3585
Realtor, Businessman, Engineer

**Comment by ThomIves** [comment]
I don't think this item should be closed yet. Your procedure is not yet working as stated. I don't mean blame by this. Rather, I am commiting to collaboration until it's done. OK?

**Comment by maaktweluit** [comment]
Hi Thom,

When it is not working for you the issue is indeed not closed yet :)

From the raw output you send here i can extract the command you should run for hyperg on another port:

Original = port 3282
```
hyperg --db /home/thom/.local/share/golem/default/mainnet/hyperg --logfile /home/thom/.local/share/golem/default/mainnet/logs/hyperg.log
```

Updated port = 3283
```
hyperg --db /home/thom/.local/share/golem/default/mainnet/hyperg --logfile /home/thom/.local/share/golem/default/mainnet/logs/hyperg.log --port 3283
```

Let me know if this works for you!

**Comment by ThomIves** [comment]
Hi maaktweluit, Still not working. Would you be willing to TeamView into my machine and see if you can get it working? If possible, I'd like to watch when you do it. Worst case, if you do it without me, I'd like to help document the procedure for our greater community. Thom

**Comment by badb** [comment]
@ThomIves Can you please join chat.golem.network and ask user dixon.hill for help with reference to this issue?

**Comment by ThomIves** [comment]
Actually, maaktweluit is helping me through a private chat on golem chat,
and we plan to publish all that we figure out from my issue. Sound OK?


Thom Ives, Call or Text 208.391.3585
Realtor, Businessman, Engineer

**Comment by AcesHigh70** [comment]
Was this process ever documented where other users can use it?

**Comment by tlaurion** [comment]
@thomives what was the actual working procedure? That would help others. Thanks.

**Comment by jet86** [comment]
Here's an explanation of how to port forward to multiple nodes I wrote a while ago. It should work in most cases, though in some cases you may need to forward to different internal ports and update the ports the golem app is listening on on each node (I don't have a write-up of how to do this currently, sorry).

https://www.reddit.com/r/GolemProject/comments/8la5rk/golem_client_updates/dzfoprw/

**Comment by tlaurion** [comment]
Applies also to #2208.

How can one modify default hyperdrive port other then from command line or modifying python script?

Use case: 

- Buy VPN service that supports port forwarding. 
- Setup VPN service's port forwarding with 3 subsequent ports (Let's say here that they are 57017 57018 and 57019)
- Create QubesOS Ubuntu template (Why not debian support, guys....)
- Setup golem VM to sit behind VPN proxyvm: sys-net <-> sys-firewall <-> VpnWithPortForwardingSetuped <-> Golem
- Apply [QubesOS port forwarding](https://gist.github.com/Joeviocoe/6c4dc0c283f6d6c5b1a3f5af8793292b) to the above networking chain:
`for i in {57017..57019}; do ./qvm-portfwd-iptables golem $i tcp; done`
- Modify `/home/user/.local/share/golem/default/mainnet/app_cfg.ini `:
```
start port = 57017
end port = 57019
rpc port = 57019
seed port = 57017
```

- Check logs:
```
INFO     golem.task.taskserver                Port 57018 opened - listening3:35 
7017, Task=57018, Hyperdrive=3282
```

- Starting hyperg manually on port to 57019:
`hyperg --db /home/user/.local/share/golem/default/mainnet/hyperg --logfile /home/user/.local/share/golem/default/mainnet/logs/hyperg.log --port 57019`
> 2018-12-10 11:17:44.033 [HyperG] [INFO] Opening HyperG log file: /home/user/.local/share/golem/default/mainnet/logs/hyperg.log
> 2018-12-10 11:17:44.039 [HyperG] [INFO] HyperG [0.2.6]
> 2018-12-10 11:17:44.059 [HyperG] [INFO] TCP listening on 0.0.0.0:57019

And then starting golemapp:
`golemapp --accept-terms --password badpassword --mainnet`

Results in:
```
[Failure instance: Traceback: <class 'twisted.internet.error.CannotListenErr
or'>: Couldn't listen on localhost:57019: [Errno 98] Address already in use.
```



....
@maaktweluit  @badb :
**Where is one supposed to configure hyperdrive to use another port then 3282-> 57019 ?**

**Comment by tlaurion** [comment]
Seems like 4 port forwards need to be setuped. (3 used by golem, 1 by hyperg)
And `hyperg --port `needs to be started prior to golemapp, and then everything is fine. Doing manually is error prone.

**Would be lovely that hyperg port would be setuped by golemapp and a config option is being presented to change it manually in configuration, like other options for golem.**

**Comment by maaktweluit** [comment]
Resolved by #4054

You can now configure your hyperg port when starting golem ( from version `0.19.2` )
